### PR TITLE
[CLI] Add feature to generate amber app from a recipe

### PR DESCRIPTION
**Pull request by damianham**
### Description of the Change

This feature adds the ability to generate amber apps from a recipe.  Only apps are generated at the moment, the next stage will be to implement scaffolding from a recipe.  The recipe to use can be either a built in recipe (from src/amber/cli/recipes), a directory, or a zip file in the repo https://github.com/AmberRecipes/recipes.  The README for the git repo explains a bit more.

### Alternate Designs
None considered
### Benefits

Users will be able to generate apps with varying configs and different, shards and shard versions, 
Users will be able to use different tooling easily e.g npm or yarn, webpack or  browserify, gulp or grunt etc.

### Possible Drawbacks
There will be a multitude of app flavours so it might be more difficult to support users as they will have a wide variety of app configuration and features.

### CODE REVIEW

**Comment by faustinoaq** [comment]
Fixes #629

File: src/amber/cli/recipes/file_entries.cr [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      abs = File.join(abs, entry)
+      if Dir.exists?(abs)
+        each_file abs, rel, &block
+      else
+        block.call abs, rel
+      end
+    end
+
+    def pack_liquid(files, abs, rel)
+      begin
```
[the remark this answers is not in the document]
Comment (damianham) [reply src/amber/cli/recipes/file_entries.cr]: Huh!   the begin block is a begin rescue end block
I wanted to catch any exceptions thrown by the Liquid template engine and print a useful message which most importantly identifies the file that failed parsing.  Without that the exception would bubble up to the top frame and just print the exception without any context information.

**Comment by robacarp** [comment]
This seems like a good idea but man it's a big pull. I'm procrastinating and wondering, would it be possible to break it into smaller sections by leaving most of the current generators as is and making pulls for each of them independently? Or is that what's already being done?

**Comment by damianham** [comment]
It doesn't change the existing generators - they are still in place and this only kicks in when a recipe is given otherwise the existing generators are used.  I am working on the other generators at the moment e.g. scaffold, model, controller for a recipe so they can be an independent PR.  This one only adds the app generator.

**Comment by eliasjpr** [comment]
> Another option would be to NOT have any recipes built in to amber since the default app recipe (which makes this such a big pull) is a duplication of the existing app template anyway. The built in default app recipe makes testing easier though as there is no need to download a recipe from github.

@damianham I agree with your comment

**Comment by damianham** [comment]
@eliasjpr yeah no problem - I just pushed a commit that removes the default recipe so it is a lot smaller now.  This is still very much a WIP - I just also changed the recipe zip file path but getting there bit by bit.

**Comment by robacarp** [comment]
@damianham I added the `wip` label to match up with your statement above about being a work in progress. Feel free to remove as desired. Thanks again for the good idea and good work. I think this could really help organize a lot of the disparate thoughts people have about generating different types of applications with Amber.

**Comment by damianham** [comment]
I think I have finalised the design of this feature so I will set out below the design decisions and how it works for your input, which will be much appreciated.

1. Recipes are contained in a zip file and the zip file is stored in the github repository https://github.com/AmberRecipes/recipes.  I felt it would be easier allow wider admin to this repository that is not the core amber github organisation and so make the admin of accepting recipe PRs easier with no risk to the core project.  I have no objection to moving the repository into the core amber github organisation.

2. Recipes are stored in a folder of the contributor's github account name.  There *may* not be that many recipe contributors so the root folder may not become over crowded with contributor folders.  Further there is nothing to stop contributor B changing a recipe from contributor A slightly and submitting the new recipe as a PR that adds a new recipe to A's folder (as long as A accepts the PR and B is mentioned as the new recipe author of course).  There is also nothing against having root folders for themes either so the root folder could contain a mix of contributor folders and theme folders.

3.  A recipe is identified by the path to the zip from the root without the '.zip' suffix.  Thus the recipe 'damianham/default' identifies the recipe at https://github.com/AmberRecipes/recipes/damianham/default.zip

4. Recipe zips can contain recipes for app, controller and scaffold components, all other components default to the existing built in template generator.  I felt that other components are unlikely to need recipe support as they should be standard.  I decided to include multiple components in 1 zip file to make it easier for recipe contributors to create recipes for all the components in one zip rather than as separate zips for each type.

5. To use a recipe to generate a component you call amber new|generate and add the recipename prefixed by '-r' as an option, e.g
amber new mynewapp -r damianham/default   #  creates an app from that recipe from github
cd mynewapp && amber g scaffold Post title:string body:text  # scaffolds posts using the same recipe.

Once an app has been created with a recipe the recipe identifier is stored in .amber.yml and is used when generating subsequent components.  I felt that if you generated the app with a recipe then you very likely want to generate the components with the same recipe so no need to specify the recipe again, and you may always forget to specify the recipe later and get the wrong code generated.

6. Downloaded recipes are cached.  I felt this would make it easier to subsequently generate more components.

7. You can supply a directory as the recipe argument and if it contains a template for the component being generated it will be used,  e.g.

amber new mynewapp -r /path/to/a/recipe  

The path '/path/to/a/recipe' should contain an 'app' template folder to generate an app, similarly for controller and scaffold.

This makes it very easy to download a recipe, unzip it, modify it to suit your tastes and use your modifications to generate stuff without submitting a PR.

8.  Any errors in finding the appropriate template for the component being generated and it falls back to the built in template generator.  This one is questionable - is it better to generate nothing and raise an error or to at least generate something (there is a message that the recipe couldn't be found).

9.  Liquid is used as the template language so this adds a shard to the project.  ECR based template binding is based on macros which are processed at compile time so didn't work with dynamic templates.  Liquid allows us to setup a context and generate dynamically at run time.

**Comment by faustinoaq** [comment]
> Recipes are contained in a zip file and the zip file is stored in the github repository https://github.com/AmberRecipes/recipes. I felt it would be easier allow wider admin to this repository that is not the core amber github organisation and so make the admin of accepting recipe PRs easier with no risk to the core project. I have no objection to moving the repository into the core amber github organisation.

I think we should move it to amberframework organization, @amberframework/contributors WDYT?

> Liquid is used as the template language so this adds a shard to the project. ECR based template binding is based on macros which are processed at compile time so didn't work with dynamic templates. Liquid allows us to setup a context and generate dynamically at run time.

Awesome!!! :clap:  :sparkles:

**Comment by damianham** [comment]
travis test run now passes on PR 710 which supercedes this PR.  In effect this PR and PR 709 can both be closed and only PR 710 merged into master as 710 contains all commits in this PR and 709 and is really the meat on the bones since it incorporates the scaffold generator.

I also agree that the recipes repository should be part of amberframework - but using an independent organisation made it easier to develop this feature.  Now the design has stabilized we can move the recipe repository to amberframework organisation (preferably before I start on the documentation for recipes).

**Comment by robacarp** [comment]
@damianham Thanks for your work. I'll close this issue in favor of #710

**Pull request closed without merging**